AndroidUnityWrapper~/unity/src/main/java/com/marigold/sdk/unity/MarigoldWrapper.kt

```diff
@@ -0,0 +1,141 @@
+import java.lang.reflect.InvocationTargetException
+
+/**
+ * Created by Affian on 10/06/15.
+ *
+ */
+object MarigoldWrapper {
+    @VisibleForTesting
+    internal var marigold = Marigold()
+    @VisibleForTesting
+    internal var messageStream = MessageStream()
+    @VisibleForTesting
+    internal var unitySender = UnitySender()
+    @VisibleForTesting
+    internal var started = false
+
+    fun start() {
+        if (started) return
+        started = true
+
+        val activity = UnityPlayer.currentActivity
+        setupMessageHandling(activity)
+
+        val broadcastManager = LocalBroadcastManager.getInstance(activity)
+        broadcastManager.registerReceiver(object : BroadcastReceiver() {
+            override fun onReceive(context: Context, intent: Intent) {
+                val count = intent.getIntExtra(MessageStream.EXTRA_UNREAD_MESSAGE_COUNT, 0)
+                unitySender.sendUnityMessage(MESSAGE_STREAM_UNITY, MESSAGE_STREAM_RECEIVE_UNREAD_COUNT, count.toString())
+            }
+        }, IntentFilter(Marigold.ACTION_MESSAGE_COUNT_UPDATE))
+
+        setWrapperInfo()
+    }
+
+    fun updateLocation(latitude: Double, longitude: Double) {
+        marigold.updateLocation(Location("Unity").apply {
+            this.latitude = latitude
+            this.longitude = longitude
+        })
+    }
+
+    fun deviceId() {
+        marigold.getDeviceId(object : Marigold.MarigoldHandler<String?> {
+            override fun onSuccess(value: String?) {
+                unitySender.sendUnityMessage(MARIGOLD_UNITY, MARIGOLD_RECEIVE_DEVICE_ID, value ?: "")
+            }
+            override fun onFailure(error: Error) {
+                unitySender.sendErrorMessage(MARIGOLD_UNITY, error)
+            }
+        })
+    }
+
+    fun logRegistrationEvent(userId: String?) {
+        marigold.logRegistrationEvent(userId)
+    }
+
+    fun setInAppNotificationsEnabled(enabled: Boolean) {
+        marigold.setInAppNotificationsEnabled(enabled)
+    }
+
+    fun setGeoIpTrackingEnabled(enabled: Boolean) {
+        marigold.setGeoIpTrackingEnabled(enabled, object : Marigold.MarigoldHandler<Void?> {
+            override fun onSuccess(value: Void?) {}
+            override fun onFailure(error: Error) {
+                unitySender.sendErrorMessage(MARIGOLD_UNITY, error)
+            }
+        })
+    }
+
+    fun requestNotificationPermission() {
+        marigold.requestNotificationPermission(UnityPlayer.currentActivity)
+    }
+
+    fun syncNotificationSettings() {
+        marigold.syncNotificationSettings()
+    }
+
+    //region Private
+
+    private fun setWrapperInfo() {
+        try {
+            val cArg: Array<Class<String>?> = arrayOfNulls(2)
+            cArg[0] = String::class.java
+            cArg[1] = String::class.java
+            val setWrapperMethod = Marigold.Companion::class.java.getDeclaredMethod("setWrapper", *cArg)
+            setWrapperMethod.isAccessible = true
```

Comment:
Yeah, I guess it's not so much about it being misused as keeping a clean API. It would feel weird to have a method in there that customers should never be using. I go back and forth on it a bit but in the end I guess I prefer a little reflection in the wrapper SDKs to negatively impacting the native SDKs.

Reply:
It's interesting I'm going back and forth myself, on rereading this. I supposed in terms of API design this is the correct choice.